internal/extender/resourcereservations.go

```diff
@@ -15,26 +15,281 @@
+}
+
+func (rrm *ResourceReservationManager) bindExecutorToSoftReservation(ctx context.Context, executor *v1.Pod, node string) error {
+	driver, err := rrm.podLister.getDriverPod(ctx, executor)
+	if err != nil {
+		return err
+	}
+	sparkResources, err := sparkResources(ctx, driver)
+	if err != nil {
+		return err
+	}
+	softReservation := v1beta1.Reservation{
+		Node:   node,
+		CPU:    sparkResources.executorResources.CPU,
+		Memory: sparkResources.executorResources.Memory,
+	}
+	return rrm.softReservationStore.AddReservationForPod(ctx, driver.Labels[common.SparkAppIDLabel], executor.Name, softReservation)
+}
+
+func (rrm *ResourceReservationManager) getUnboundReservations(ctx context.Context, executor *v1.Pod) (map[string]string, error) {
+	resourceReservation, ok := rrm.GetResourceReservation(executor)
+	if !ok {
+		return nil, werror.ErrorWithContextParams(ctx, "failed to get resource reservation")
+	}
+	activePodNames, err := rrm.getActivePodNames(ctx, executor)
+	if err != nil {
+		return nil, err
+	}
+
+	unboundReservationsToNodes := make(map[string]string, len(resourceReservation.Spec.Reservations))
+	for reservationName, reservation := range resourceReservation.Spec.Reservations {
+		podIdentifier, ok := resourceReservation.Status.Pods[reservationName]
+		if !ok || !activePodNames[podIdentifier] {
+			unboundReservationsToNodes[reservationName] = reservation.Node
+		}
+	}
+	return unboundReservationsToNodes, nil
+}
+
+func (rrm *ResourceReservationManager) getFreeExtraExecutorSpots(ctx context.Context, executor *v1.Pod) (int, error) {
+	extraExecutorCount := 0
+	sr, ok := rrm.softReservationStore.GetSoftReservation(executor.Labels[common.SparkAppIDLabel])
+	if !ok {
+		return 0, nil
+	}
+	extraExecutorCount = len(sr.Reservations)
+	driver, err := rrm.podLister.getDriverPod(ctx, executor)
+	if err != nil {
+		return 0, err
+	}
+	sparkResources, err := sparkResources(ctx, driver)
+	if err != nil {
+		return 0, err
+	}
+	maxAllowedExtraExecutors := sparkResources.maxExecutorCount - sparkResources.minExecutorCount
+	return int(math.Max(float64(maxAllowedExtraExecutors-extraExecutorCount), 0)), nil
+}
+
+// getActivePodNames returns a map of pod names that are still active in the passed pod's namespace
+func (rrm *ResourceReservationManager) getActivePodNames(ctx context.Context, pod *v1.Pod) (map[string]bool, error) {
+	selector := labels.Set(map[string]string{common.SparkAppIDLabel: pod.Labels[common.SparkAppIDLabel]}).AsSelector()
+	pods, err := rrm.podLister.Pods(pod.Namespace).List(selector)
+	if err != nil {
+		return nil, werror.WrapWithContextParams(ctx, err, "failed to list pods")
+	}
+	activePodNames := make(map[string]bool, len(pods))
+	for _, pod := range pods {
+		if !rrm.IsPodTerminated(pod) {
+			activePodNames[pod.Name] = true
+		}
+	}
+	return activePodNames, nil
+}
+
+// IsPodTerminated returns whether the pod is considered to be terminated
+func (rrm *ResourceReservationManager) IsPodTerminated(pod *v1.Pod) bool {
```

Comment:
The following three methods (`IsPodTerminated`, `NewResourceReservation` and `ExecutorReservationName`) should not be exported. Unfortunately, they're being used in failover and refactoring failover to use `ResourceReservationManager` is out of the scope of this PR. 
We should consider limiting them back as soon as we refactor failover.

Reply:
The initial intent was to prepare for moving all reservation creations to go through `ResourceReservationManager` which is why I made those private methods to this. 
We can do this at a later date though when refactoring failover, reverted for now.